### PR TITLE
Add serializers for data types openedDataObjInp_t and bytesBuf_t

### PR DESCRIPTION
**Pull request by ilarik**
Hi,

As discussed previously, here's a pull request with the serializer for openedDataObjInp_t.

- ilari

### CODE REVIEW

**Comment by trel** [comment]
Thanks Ilari - please rebase this so we have a clean look at the new code.

Do we have/need a test for a serializer?

File: server/re/src/irods_re_serialization.cpp [hunk 1]
```diff
@@ -969,6 +969,63 @@ namespace irods {
...
+		    _out["l1descInx"] = std::to_string(ptr->l1descInx);
+		    _out["len"] = std::to_string(ptr->len);
+		    _out["whence"] = std::to_string(ptr->whence);
+		    _out["oprType"] = std::to_string(ptr->oprType);
+		    _out["offset"] = std::to_string(ptr->offset);
+		    _out["bytesWritten"] = std::to_string(ptr->bytesWritten);
+		    
+		    serialize_parameter(ptr->condInput, _out);
+		}
+
```
Comment (korydraughn) [inline server/re/src/irods_re_serialization.cpp:990]: remove extra line

File: server/re/src/irods_re_serialization.cpp [hunk 2]
```diff
@@ -969,6 +969,63 @@ namespace irods {
...
+							       serialized_parameter_t& _out) 
+	{
+	    try {
+		openedDataObjInp_t **ptr = boost::any_cast<openedDataObjInp_t**>(_p);
+		
+		// if we can, dereference and serialize
+		if (ptr && *ptr) {
+		    serialize_openedDataObjInp_ptr(*ptr, _out);
+		}
+
```
Comment (korydraughn) [inline server/re/src/irods_re_serialization.cpp:1015]: remove extra line

File: server/re/src/irods_re_serialization.cpp [hunk 3]
```diff
@@ -1019,7 +1076,9 @@ namespace irods {
                 { std::type_index(typeid(rodsObjStat_t*)), serialize_rodsObjStat_ptr },
                 { std::type_index(typeid(genQueryInp_t*)), serialize_genQueryInp_ptr },
                 { std::type_index(typeid(genQueryOut_t*)), serialize_genQueryOut_ptr },
-                { std::type_index(typeid(char**)), serialize_char_ptr_ptr }
+                { std::type_index(typeid(char**)), serialize_char_ptr_ptr },
+		{ std::type_index(typeid(openedDataObjInp_t*)), serialize_openedDataObjInp_ptr },
+		{ std::type_index(typeid(openedDataObjInp_t**)), serialize_openedDataObjInp_ptr_ptr }
```
Comment (korydraughn) [inline server/re/src/irods_re_serialization.cpp:1081]: indentation

File: server/re/src/irods_re_serialization.cpp [hunk 4]
```diff
@@ -969,6 +969,63 @@ namespace irods {
...
+
+	static irods::error serialize_openedDataObjInp_ptr(boost::any _p,
+							   serialized_parameter_t& _out) {
+	    try {
+		openedDataObjInp_t *ptr = boost::any_cast<openedDataObjInp_t*>(_p);
+
+		// for a valid ptr, we serialize the immediate properties + keyValPair
+		if (ptr) {
+		    _out["l1descInx"] = std::to_string(ptr->l1descInx);
+		    _out["len"] = std::to_string(ptr->len);
```
Comment (korydraughn) [inline server/re/src/irods_re_serialization.cpp:982]: @trel how do we feel about using all of the letters here?

**Review by korydraughn — COMMENTED** [review]
Looks good to me. I think a test is needed. The following can be used as a reference. Maybe don't use "**if**" in the test name :)

https://github.com/irods/irods/blob/11c303f29cde72183dd6e8106f4c534ff5dea3d9/scripts/irods/test/test_dynamic_peps.py#L28

File: server/re/src/irods_re_serialization.cpp (hunk 4 above)
Comment (trel) [reply server/re/src/irods_re_serialization.cpp:982]: Big fan of using the letters.

But on the other hand, this reflects the existing data structure itself... so, unless we're expanding all the rest of them, I think we just leave it.   We can use more letters when we write wrappers or a new thing.

File: server/re/src/irods_re_serialization.cpp (hunk 1 above)
Comment (ilarik) [reply server/re/src/irods_re_serialization.cpp:990]: done

File: server/re/src/irods_re_serialization.cpp (hunk 2 above)
Comment (ilarik) [reply server/re/src/irods_re_serialization.cpp:1015]: done

File: server/re/src/irods_re_serialization.cpp (hunk 3 above)
Comment (ilarik) [reply server/re/src/irods_re_serialization.cpp:1081]: fixed emacs indentation for spaces instead of tabs, done.

**Comment by ilarik** [comment]
Commenting to point out that this pull request was renamed to reflect the addition of another serializer (bytesBuf_t), which is needed in PEPs added in 4-2-stable while this was in progress.

File: server/re/src/irods_re_serialization.cpp [hunk 5]
```diff
@@ -969,6 +970,98 @@ namespace irods {
...
+                }
+            }
+            catch (std::exception&) {
+                return ERROR(
+                    INVALID_ANY_CAST,
+                    "failed to cast bytesBuf_t ptr");
+            }
+
+            return SUCCESS();
+        } // serialize_bytesBuf_t
```
Comment (trel) [inline server/re/src/irods_re_serialization.cpp:1063]: serialize_bytesBuf_ptr

**Comment by trel** [comment]
i think this is ready for squashing and referencing 5408 by name in the commit message (all without the # sign).

once it looks really solid, you can add the # to the commit message and we'll merge.

then a single cherry pick for master in a new PR

File: server/re/src/irods_re_serialization.cpp (hunk 5 above)
Comment (ilarik) [reply server/re/src/irods_re_serialization.cpp]: whoops... fixed and committed.

**Comment by alanking** [comment]
Have you considered adding a test (as @korydraughn suggested here: https://github.com/irods/irods/pull/5417#pullrequestreview-592733000)? Otherwise, everything looks fine to me.

**Comment by ilarik** [comment]
> Have you considered adding a test (as @korydraughn suggested here: [#5417 (review)](https://github.com/irods/irods/pull/5417#pullrequestreview-592733000))? Otherwise, everything looks fine to me.

Probably there should be test(s). I don't know how your testing framework works.

**Comment by trel** [comment]
added a test here...  https://github.com/irods/irods/compare/4-2-stable...trel:pr-2021-02-17

jenkins not behaving for me locally, @alanking can you build and run the one test `test_openedDataObjInp_t_and_bytesBuf_t_serializers__issue_5408`?

**Comment by ilarik** [comment]
> added a test here... [4-2-stable...trel:pr-2021-02-17](https://github.com/irods/irods/compare/4-2-stable...trel:pr-2021-02-17)
> 
> jenkins not behaving for me locally, @alanking can you build and run the one test `test_openedDataObjInp_t_and_bytesBuf_t_serializers__issue_5408`?

Great, thanks a lot! It would have taken some time for me to figure out at first.

**Comment by trel** [comment]
those were done blind - please eyeball for valid claims being made.

**Comment by alanking** [comment]
I'm working on shaping up the tests.

**Comment by alanking** [comment]
Got the python tests to a pretty good spot here: https://github.com/alanking/irods/commit/a0ae6d9ab867e33cea735e4e04afb704054b1a60

And added some unit tests for fun: https://github.com/alanking/irods/commit/9442caee2c91796ef3db9cf806af29fd42fc8c82

Built atop the @trel branch: https://github.com/alanking/irods/compare/trel:pr-2021-02-17..pr-2021-02-17

**Pull request merged**